# Surplus Sales Exploratory Data Analysis 

In this notebook, we will look at the sale of Surplus from the dataset called '2.28 Sold Lots.xlsx' which shows all the sales of Surplus store up to 02/28/2024. 

In [2]:
import pandas as pd
import numpy as np
surplus = pd.read_excel('2.28 Sold Lots.xlsx')
surplus

Record ID Warehouse Location  \
0      237067 || TABLE: 60" X 30" LIGHT WDGRN TOP ROL...                NaN   
1        237025 || Freight Shipping Fees for Lot #229195                NaN   
2      236999 || Freight Fee for Lots #235920, #22458...                NaN   
3      236986 || BARREL: 55 GAL CHEMICAL STORAGE CONT...                NaN   
4              236982 || Gov Deals Fee for Lot #229195.0                NaN   
...                                                  ...                ...   
20374         207865 (SSIS ID: 136767) || POSITIONER: XY                NaN   
20375  207864 (SSIS ID: 137162) || RESERVOIR: GLASS S...                NaN   
20376  207863 (SSIS ID: 117604) || SURGICAL TOOL KIT:...                NaN   
20377  207862 (SSIS ID: 135859) || TRAYS: FOR DISSECT...                NaN   
20378  207861 (SSIS ID: 133399) || PROJECTOR SCREEN: ...                NaN   

                                            Lot Category          Lot Status  \
0                                              Furniture  Sold Not Deposited   
1      Shipping & Handling (Not Viewed on Public Faci...  Sold Not Deposited   
2      Shipping & Handling (Not Viewed on Public Faci...  Sold Not Deposited   
3                             Laboratory/Research Equip.  Sold Not Deposited   
4      Gov. Deals & Public Surplus Commission Rates ...  Sold Not Deposited   
...                                                  ...                 ...   
20374                         Laboratory/Research Equip.           Journaled   
20375                                          Glassware           Journaled   
20376                            Clinical/Medical Equip.           Journaled   
20377                         Laboratory/Research Equip.           Journaled   
20378                                Audio/Visual Equip.           Journaled   

      Journal Type     Sale Surplus Transfer Request  \
0              NaN  52810.0                      NaN   
1              NaN  52814.0                      NaN   
2              NaN  52480.0                      NaN   
3              NaN  52823.0                      NaN   
4              NaN  52814.0                      NaN   
...            ...      ...                      ...   
20374      Type 20  44334.0             STR #: 10992   
20375      Type 20  45772.0             STR #: 10953   
20376      Type 16  43208.0             STR #: 10125   
20377      Type 20  43273.0              STR #: 3824   
20378      Type 20  45877.0              STR #: 3761   

       Selling Dept's Financial String (at time of journaling) Delivery Type  \
0                                           1.000664e+10                 NaN   
1                                           1.000664e+10                 NaN   
2                                           1.000664e+10                 NaN   
3                                           1.000664e+10                 NaN   
4                                           1.000664e+10                 NaN   
...                                                  ...                 ...   
20374                                       1.023844e+10                 NaN   
20375                                       1.000210e+10                 NaN   
20376                                       1.500952e+10                 NaN   
20377                                       1.000709e+10                 NaN   
20378                                       1.003860e+10                 NaN   

        Lot #  ...                                   Seller Fin Unit  \
0      237067  ...               5000141 || Surplus Sales Operations   
1      237025  ...               5000141 || Surplus Sales Operations   
2      236999  ...               5000141 || Surplus Sales Operations   
3      236986  ...               5000141 || Surplus Sales Operations   
4      236982  ...               5000141 || Surplus Sales Operations   
...       ...  ...                                               ...   

## Data Cleaning 

In [3]:
#Basic info about the data 
surplus.shape

(20379, 144)

The code above shows us the dimensionality of the dataframe. We can see that there are 20,379 rows (20,379 entries) and 144 columns (144 attributes for each transaciton). We can easily infer that the dataset is extremely big. Moreover, from first glance of the data, we can see that a lot data are missing we will clean this data to be more concise and usable. 

First, we will drop all the columns that are empty with the code below. 

In [4]:
sp = surplus.dropna(how='all', axis=1)

In [5]:
sp.shape
#Shows us that 10 columns were completely empty. 

(20379, 134)

In [6]:
#Sorting the rows by latest transaction date
sp = sp.sort_values(by = 'Transaction Date', ascending = False)

We then sort the data by transaction date, starting from the newest transaction. We used 'Transaction Date' as since out of all the dates for each transaction, it seems the most simpliest with most applicable format. 

Still, we have 134 columns of attributes per transaction. It is still obvious that some columns are just mostly NaN or empty. We will drop all the columns that are more than 60% empty. This will allow us to actually run some test and get significant results rather than just null values. 

In [7]:
#Now dropping columns that are more than 60% empty
sp = sp.loc[:, (sp.isin([' ','NULL',0]) | sp.isnull()).mean() <= .6]

In [8]:
sp.shape

(20379, 84)

This is what the dataframe looks like so far. It is much more concise and holds meaning data now. 

In [9]:
sp

Record ID  \
1        237025 || Freight Shipping Fees for Lot #229195   
3831             229195 || MIXING CONSOLE: AUDIO DIGITAL   
331         236006 || BOTTLES: SAMPLE, 2L WITH SCREW CAP   
11     236963 || MATRESS; CLININCAL; FOR GURNEY 26 X ...   
10     236964 || MATRESS; CLININCAL; FOR GURNEY 30 X ...   
...                                                  ...   
18733  210523 (SSIS ID: 114678) || TRAY- STAINLESS STEEL   
17784  211994 (SSIS ID: 138047) || JAR: LARGE, GLASS ...   
17785  211993 (SSIS ID: 138046) || JAR: SMALL, GLASS ...   
18273  211188 (SSIS ID: 130136) || DECORATIVE TREE- F...   
568        235615 || FUNCTION GENERATOR: ELECTRONIC TEST   

                                            Lot Category          Lot Status  \
1      Shipping & Handling (Not Viewed on Public Faci...  Sold Not Deposited   
3831                                 Audio/Visual Equip.  Sold Not Deposited   
331                                            Glassware  Sold Not Deposited   
11                               Clinical/Medical Equip.  Sold Not Deposited   
10                               Clinical/Medical Equip.  Sold Not Deposited   
...                                                  ...                 ...   
18733                         Laboratory/Research Equip.           Journaled   
17784                                          Glassware           Journaled   
17785                                          Glassware           Journaled   
18273                                          Furniture           Journaled   
568                               Electronic Test Equip.  Sold and Deposited   

      Journal Type     Sale Surplus Transfer Request  \
1              NaN  52814.0                      NaN   
3831           NaN  52814.0             STR #: 44247   
331            NaN  52823.0             STR #: 46640   
11             NaN  52824.0             STR #: 47201   
10             NaN  52824.0             STR #: 47201   
...            ...      ...                      ...   
18733      Type 20  41119.0             STR #: 10155   
17784      Type 20  41120.0             STR #: 37545   
17785      Type 20  41119.0             STR #: 37545   
18273      Type 20  41115.0             STR #: 10155   
568        Type 20  52736.0             STR #: 46483   

       Selling Dept's Financial String (at time of journaling)   Lot #  \
1                                           1.000664e+10        237025   
3831                                        1.009610e+10        229195   
331                                         1.000044e+10        236006   
11                                          1.501288e+10        236963   
10                                          1.501288e+10        236964   
...                                                  ...           ...   
18733                                       1.000664e+10        210523   
17784                                       1.023585e+10        211994   
17785                                       1.023585e+10        211993   
18273                                       1.000664e+10        211188   
568                                         1.001132e+10        235615   

       Number of Journal Lines             QR Code - Image  ...  \
1                            0  QR Code - Lot #:237025.png  ...   
3831                         0  QR Code - Lot #:229195.png  ...   
331                          0  QR Code - Lot #:236006.png  ...   
11                           0  QR Code - Lot #:236963.png  ...   
10                           0  QR Code - Lot #:236964.png  ...   
...                        ...                         ...  ...   
18733                        4                         NaN  ...   
17784                        4                         NaN  ...   
17785                        4                         NaN  ...   
18273                        4                         NaN  ...   
568                          0  QR Code - Lot #:235615.png  ...   

Now, we can see that there are duplicate columns with different column name. Although this might've been important in termns logging all the information about the transaction, in our case, this is just duplicate data that is not useful to us. For example, look at the toy dataframe below. 

In [10]:
toy  = sp[['Lot Category', 'Category']]
toy

Lot Category  \
1      Shipping & Handling (Not Viewed on Public Faci...   
3831                                 Audio/Visual Equip.   
331                                            Glassware   
11                               Clinical/Medical Equip.   
10                               Clinical/Medical Equip.   
...                                                  ...   
18733                         Laboratory/Research Equip.   
17784                                          Glassware   
17785                                          Glassware   
18273                                          Furniture   
568                               Electronic Test Equip.   

                                                Category  
1      Shipping & Handling (Not Viewed on Public Faci...  
3831                                 Audio/Visual Equip.  
331                                            Glassware  
11                               Clinical/Medical Equip.  
10                               Clinical/Medical Equip.  
...                                                  ...  
18733                         Laboratory/Research Equip.  
17784                                          Glassware  
17785                                          Glassware  
18273                                          Furniture  
568                               Electronic Test Equip.  

[20379 rows x 2 columns]

Some columns hold the same exact data that is not necessary. Now, we will remove duplicate columns of data, leaving only the first occuring column. 

In [11]:
#Now lets remove duplicate columns of data 
sp = sp.T.drop_duplicates().T

In [12]:
sp.shape

(20379, 77)

### Columns with String (Text) Data 

For more in-depth look cleaning of the data, we will only look at the columns with the string data right now. 

In [22]:
sp_string = sp.loc[:, sp.iloc[0].map(type) == str]

In [25]:
sp_string.shape
#From this we can see that out of the 77 columns left after removing columns that are empty or 60% empty,
#32 of the column is string data

(20379, 32)

In [26]:
sp_string

Record ID  \
1        237025 || Freight Shipping Fees for Lot #229195   
3831             229195 || MIXING CONSOLE: AUDIO DIGITAL   
331         236006 || BOTTLES: SAMPLE, 2L WITH SCREW CAP   
11     236963 || MATRESS; CLININCAL; FOR GURNEY 26 X ...   
10     236964 || MATRESS; CLININCAL; FOR GURNEY 30 X ...   
...                                                  ...   
18733  210523 (SSIS ID: 114678) || TRAY- STAINLESS STEEL   
17784  211994 (SSIS ID: 138047) || JAR: LARGE, GLASS ...   
17785  211993 (SSIS ID: 138046) || JAR: SMALL, GLASS ...   
18273  211188 (SSIS ID: 130136) || DECORATIVE TREE- F...   
568        235615 || FUNCTION GENERATOR: ELECTRONIC TEST   

                                            Lot Category          Lot Status  \
1      Shipping & Handling (Not Viewed on Public Faci...  Sold Not Deposited   
3831                                 Audio/Visual Equip.  Sold Not Deposited   
331                                            Glassware  Sold Not Deposited   
11                               Clinical/Medical Equip.  Sold Not Deposited   
10                               Clinical/Medical Equip.  Sold Not Deposited   
...                                                  ...                 ...   
18733                         Laboratory/Research Equip.           Journaled   
17784                                          Glassware           Journaled   
17785                                          Glassware           Journaled   
18273                                          Furniture           Journaled   
568                               Electronic Test Equip.  Sold and Deposited   

                  QR Code - Image Added to Journal  \
1      QR Code - Lot #:237025.png          Not Yet   
3831   QR Code - Lot #:229195.png          Not Yet   
331    QR Code - Lot #:236006.png          Not Yet   
11     QR Code - Lot #:236963.png          Not Yet   
10     QR Code - Lot #:236964.png          Not Yet   
...                           ...              ...   
18733                         NaN            Added   
17784                         NaN            Added   
17785                         NaN            Added   
18273                         NaN            Added   
568    QR Code - Lot #:235615.png          Not Yet   

      Seller Business Org Name (for Journal)  \
1                                UCSD CAMPUS   
3831                             UCSD CAMPUS   
331                              UCSD CAMPUS   
11                       UCSD MEDICAL CENTER   
10                       UCSD MEDICAL CENTER   
...                                      ...   
18733                            UCSD CAMPUS   
17784                            UCSD CAMPUS   
17785                            UCSD CAMPUS   
18273                            UCSD CAMPUS   
568                              UCSD CAMPUS   

                     Subtitle (Rainworx) Active UCID [OLD] Active UCID  \
1                                     ,           Inactive    Inactive   
3831           YAMAHA, M7CL-32 VERSION 3          Inactive    Inactive   
331                          PYREX, 1395          Inactive    Inactive   
11                      MEDMATRESS.COM,           Inactive    Inactive   
10     STRYKER, ISOFLEX SE  1806-034-300          Inactive    Inactive   
...                                  ...               ...         ...   
18733                                 ,           Inactive    Inactive   
17784                                 ,           Inactive    Inactive   
17785                                 ,           Inactive    Inactive   
18273                                 ,           Inactive    Inactive   
568                 GW INSTEK, GFG-8020H          Inactive    Inactive   

                     Surplus Property  ...  \
1      ✓ - This is Surplus Property  ...   
3831                              NaN  ...   
331                               NaN  ...   
11                                NaN  ...   
10                  